### PR TITLE
Add age support, second attempt

### LINKED ISSUE #53: Document `age` usage
Sops [recently got `age` support](https://github.com/mozilla/sops/pull/688). Is it possible to use `age` with sops-nix?

### PR DESCRIPTION
**Pull request by dasJ**
This time with less indentation changes but with more tests and ssh keys support

Closes #53 
Closes #95

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,6 +29,16 @@ key management APIs such as AWS KMS, GCP KMS, Azure Key Vault or Hashicorp's vau
...
 
+## Supported encryption methods
+
+sops-nix supports two basic ways of encryption, gnupg and age. Gnupg is based
+on gnupg (duh) and encrypts against gnupg public keys. Private gnupg keys may
+be used to decrypt the secrets on the target machine. The tool `ssh-to-pgp` can
+be used to derive a gnupg key from a ssh (host) key in RSA format.
+
+The other method is age which is based on [age](https://github.com/FiloSottile/age).
+A tool is provided with sops-nix that can convert ssh host or user keys to age keys.
```
Comment (Mic92) [inline README.md:40]:
```suggestion
A tool is provided with sops-nix that can convert ssh host or user keys in ed25519 format to age keys.
```

File: README.md [hunk 2]
```diff
@@ -205,11 +236,20 @@ creation_rules:
 
 If you prefer having a separate GnuPG key, see [Use with GnuPG instead of ssh keys](#use-with-gnupg-instead-of-ssh-keys).
 
+### 3b. Get a age Public key for your machine
+
+The `sops-ssh-to-age` tool is used to convert any ssh public key to the age format.
+This way you can convert any key:
+```console
+$ nix run -f default.nix sops-ssh-to-age -c sh -c 'ssh-keyscan my-server.com | sops-ssh-to-age'
```
Comment (Mic92) [inline README.md:244]: Should I quickly package `sops-ssh-to-age` in nixpkgs. I did so for the other tool as well.

File: README.md [hunk 3]
```diff
@@ -364,6 +404,8 @@ If you derived your server public key from ssh, all you need in your configurati
   # sops.defaultSopsFile = "/root/.sops/secrets.yaml";
   sops.defaultSopsFile = ./secrets.yaml;
   sops.secrets.example-key = {};
+  # This is using ssh keys in the age format:
+  sops.age.sshKeyPaths = [ "/etc/ssh/ssh_host_ed25519_key" ];
```
Comment (Mic92) [inline README.md:408]: Should we just import this by default? Should be quite cheap to do so and than users cannot forget.

File: modules/sops/default.nix [hunk 4]
```diff
@@ -168,9 +213,18 @@ in {
...
     system.activationScripts.setup-secrets = let
       sops-install-secrets = (pkgs.callPackage ../.. {}).sops-install-secrets;
-    in stringAfter [ "specialfs" "users" "groups" ] ''
+    in stringAfter ([ "specialfs" "users" "groups" ] ++ optional cfg.age.generateKey "generate-age-key") ''
       echo setting up secrets...
-      ${optionalString (cfg.gnupgHome != null) "SOPS_GPG_EXEC=${pkgs.gnupg}/bin/gpg"} ${sops-install-secrets}/bin/sops-install-secrets ${checkedManifest}
+      ${optionalString (cfg.gnupg.home != null) "SOPS_GPG_EXEC=${pkgs.gnupg}/bin/gpg"} ${sops-install-secrets}/bin/sops-install-secrets ${checkedManifest}
     '';
+
+    system.activationScripts.generate-age-key = (mkIf cfg.age.generateKey) (stringAfter [] ''
```
Comment (Mic92) [inline modules/sops/default.nix:221]: `TODO`: I don't think this option is documented yet in Readme.

File: pkgs/sops-ssh-to-age/main.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return p.BytesMontgomery(), nil
+}
+
+func main() {
+	if len(os.Args) != 1 {
+		println("Usage: " + os.Args[0])
+		println("Pipe a SSH public key or the output of ssh-keyscan into it")
```
Comment (Mic92) [inline pkgs/sops-ssh-to-age/main.go:29]: Looks good. Should the Readme also mention ssh public key support?

File: modules/sops/default.nix [hunk 6]
```diff
@@ -130,30 +133,72 @@ in {
...
+        description = ''
+          Whether or not to generate the age key. If this
+          option is set to false, the key must already be
+          present at the specified location.
+        '';
+      };
+
+      sshKeyPaths = mkOption {
+        type = types.listOf types.path;
+        default = []; # If we set this like the gnupg option, we would use age by default
```
Comment (Mic92) [inline modules/sops/default.nix:159]:
```suggestion
        default = []; # If we set this like the gnupg option, we would use ed25519 by default
```

**Comment by Mic92** [comment]
Ok. My only question is if sops-nix should just import all rsa ssh host keys as gpg keys and all ed25519 ssh host keys as age keys by default? Then we would only need one ssh key option which would make the interface simpler and most people would not need to change anything about it. I might go over the README myself and maybe separate both parts a bit and also document how use sops age keys as a user.

File: README.md (hunk 2 above)
Comment (dasJ) [reply README.md:244]: Yeah, why not

**Comment by dasJ** [comment]
I added a new commit to address the requested changes except the one about the ssh keys option.
The main problem I see here is that we don't know the ssh key's type at eval time so we can't just use the same option.

Possible solution:
The interface would be required to add some option like `sops.age.enable` to distinguish between the encryption types.
Then we could default `sops.age.sshKeyPaths` to the same as `sops.gnupg.sshKeyPaths` but with the ed25519 keys instead of rsa.

**Comment by dasJ** [comment]
I also added a converter for private ssh keys to age now and updated the README

**Comment by Mic92** [comment]
I moved your converters to a new project so that we can package them in nixpkgs just like ssh-to-pgp: https://github.com/Mic92/ssh-to-age
This also includes the conversion functions itself: https://github.com/Mic92/ssh-to-age/blob/main/convert.go
Could you migrate the code towards this?

File: README.md (hunk 2 above)
Comment (Mic92) [reply README.md]: Done in https://github.com/NixOS/nixpkgs/pull/136098

File: README.md [hunk 7]
```diff
@@ -173,7 +188,32 @@ uid           [ unknown] root <root@localhost>
...
+```console
+$ nix run -f default.nix ssh-pubkey-to-age -c sh -c 'ssh-add -L | ssh-pubkey-to-age'
+```
+
+Ssh public key files may also be piped into the `ssh-pubkey-to-age` tool.
+
+Finally, you need to convert your private key to the age format:
+```console
+$ mkdir -p ~/.config/sops
+$ nix run -f default.nix ssh-privkey-to-age -c ssh-privkey-to-age ~/.ssh/id_ed25519 > ~/.config/sops/keys.txt
```
Comment (tomberek) [inline README.md:213]:
```suggestion
$ nix run -f default.nix ssh-privkey-to-age -c ssh-privkey-to-age ~/.ssh/id_ed25519 > ~/.config/sops/age/keys.txt
```

File: modules/sops/default.nix (hunk 4 above)
Comment (dasJ) [reply modules/sops/default.nix:216]: Looks like I rebased that in at some point

File: README.md [hunk 8]
```diff
@@ -175,7 +190,32 @@ uid           [ unknown] root <root@localhost>
...
+```console
+$ nix-shell -p ssh-pubkey-to-age --run "ssh-add -L | ssh-pubkey-to-age"
+```
+
+Ssh public key files may also be piped into the `ssh-pubkey-to-age` tool.
+
+Finally, you need to convert your private key to the age format:
+```console
+$ mkdir -p ~/.config/sops
+$ nix-shell -p ssh-privkey-to-age --run "ssh-privkey-to-age ~/.ssh/id_ed25519 > ~/.config/sops/age/keys.txt"
```
Comment (Mic92) [inline README.md:215]: So it looks like if documentation is updated this is ready for a test drive?

**Comment by Mic92** [comment]
Thanks a lot!

**Pull request merged**